### PR TITLE
[torch.export] Warn when batch_size=1 with dynamic_shapes may cause specialization

### PR DESCRIPTION
**Pull request by gaoren002**
# [torch.export] Warn when batch_size=1 with dynamic_shapes may cause specialization

## Summary

Adds a warning in `torch.export.export()` when an example input has `batch_size=1` combined with `dynamic_shapes` specification. This addresses the 0/1 Specialization Problem where dynamic dimensions may be incorrectly specialized to static values.

## Motivation

**The Problem**: When using `torch.export.export()` with:
- An example input where `batch_size=1`
- `dynamic_shapes` parameter specifying dynamic batch dimension

The batch dimension may be specialized to constant `1` instead of remaining dynamic, breaking downstream use cases that require variable batch sizes (e.g., TensorRT deployment, ONNX runtime).

**Impact**: This affects:
- Direct users of `torch.export.export()`
- Users of `torch.onnx.export(dynamo=True)` (which calls `torch.export` internally)
- Any tools or frameworks built on top of `torch.export`

**Why torch.export?**: The root cause is in `torch.export`'s "0/1 Specialization Problem" - for technical reasons, `torch.export` assumes dynamic dimensions are not 0 or 1. By adding the warning here, we catch the issue at its source rather than in downstream libraries.

**Related Issues**: #165259, #133653, #170172

### Modified Files

1. **torch/export/__init__.py**
   - Added batch=1 check in `export()` function before calling `_export()`
   - Emits `UserWarning` when risky configuration detected
   - Simple check: ~10 lines of code

2. **test/export/test_export.py**
   - Added test case `test_warn_batch_one_dynamic_shapes`
   - Verifies warning is emitted correctly
   - Uses simple model with `Dim("batch")` dynamic shape

### Warning Message

```python
UserWarning:
Using batch_size=1 with dynamic_shapes may cause dynamic dimensions to be specialized to static.
Consider using batch_size >= 2. See: https://github.com/pytorch/pytorch/issues/165259
```

## Implementation Simplicity

**Total code change**: Only ~10 lines added to `torch.export.export()`:

```python
# Warn if batch=1 with dynamic_shapes (issue #165259, #133653, #170172)
if dynamic_shapes is not None:
    first_arg = args[0] if args else None
    if isinstance(first_arg, torch.Tensor) and first_arg.size(0) == 1:
        import warnings
        warnings.warn(
            "Using batch_size=1 with dynamic_shapes may cause dynamic dimensions to be specialized to static. "
            "Consider using batch_size >= 2. See: https://github.com/pytorch/pytorch/issues/165259",
            UserWarning,
            stacklevel=2,
        )
```

- No complex logic checking output shapes or axis specifications
- No new helper functions
- Just a simple validation before calling `_export()`

### New Test Case

```python
def test_warn_batch_one_dynamic_shapes(self):
    """Test warning when batch=1 with dynamic_shapes"""
    import warnings
    from torch.export import Dim

    class SimpleModel(torch.nn.Module):
        def forward(self, x):
            return x + 1

    model = SimpleModel()
    dummy = torch.randn(1, 16)  # batch=1

    batch = Dim("batch")
    dynamic_shapes = {"x": {0: batch}}

    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        export(model, (dummy,), dynamic_shapes=dynamic_shapes)

        # Should emit warning about batch_size=1
        self.assertGreater(len(w), 0)
        warning_message = str(w[0].message)
        self.assertIn("batch_size=1", warning_message)
```

### Existing Tests

All existing tests pass without modification.

## Breaking Changes

**None**. This PR only adds a warning message, no behavior changes.

## Example Usage

**Before** (no warning, confusing behavior):
```python
from torch.export import export, Dim

model = SimpleModel()
dummy = torch.randn(1, 16)  # batch=1
batch = Dim("batch")

ep = export(model, (dummy,), dynamic_shapes={"x": {0: batch}})
# Silently specializes batch dim to 1!
```

**After** (clear warning):
```python
ep = export(model, (dummy,), dynamic_shapes={"x": {0: batch}})
# UserWarning: Using batch_size=1 with dynamic_shapes may cause...
# Recommendation: Consider using batch_size >= 2.
```

**Fix** (user can easily resolve):
```python
dummy = torch.randn(2, 16)  # batch >= 2
ep = export(model, (dummy,), dynamic_shapes={"x": {0: batch}})
# Works correctly, batch dim remains dynamic
```

## Benefits of this Approach

1. **Catches the problem at the source** - `torch.export` is where the specialization happens
2. **Helps all users** - Not just ONNX, but anyone using `torch.export`
3. **Minimal code** - Only ~10 lines, easy to maintain
4. **No complexity** - Simple validation, no intricate logic
5. **Educational** - Points users to the root issue with clear guidance

## Response to Previous Feedback

This implementation directly addresses the maintainer's concern about complexity:
- ❌ **Previous version**: Complex logic checking output names, axes specifications, etc.
- ✅ **This version**: Simple 10-line check, just validates the core issue

The warning is now at the appropriate layer (`torch.export`) rather than a downstream consumer (`torch.onnx`).

### CODE REVIEW

**Review by justinchuby — COMMENTED** [review]
Thanks. I would accept this PR if the logic were simpler. Right now it is too complicated and error prone for the benefit it brings. I also think the warning is better suited for torch.export. @tugsbayasgalan @angelayi what do you think?
